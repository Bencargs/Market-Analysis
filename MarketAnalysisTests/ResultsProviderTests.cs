using System;
using System.Collections.Concurrent;
using System.IO;
using System.Linq;
using System.Threading.Tasks;
using ApprovalTests;
using ApprovalTests.Namers;
using ApprovalTests.Reporters;
using MarketAnalysis.Caching;
using MarketAnalysis.Factories;
using MarketAnalysis.Models;
using MarketAnalysis.Providers;
using MarketAnalysis.Repositories;
using MarketAnalysis.Services;
using MarketAnalysis.Strategy;
using MarketAnalysis.Strategy.Parameters;
using Moq;
using Newtonsoft.Json;
using NUnit.Framework;

namespace MarketAnalysisTests
{
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("ChartTests")]
    public class ResultsProviderTests : TestHarness
    {
        private IResultsProvider _target;

        [SetUp]
        public void Setup()
        {
            var marketData = CreateMarketData();
            var marketDataCache = CreateMarketDataCache(marketData);

            var investor = new Investor { DailyFunds = 10, OrderDelayDays = 3 };
            var investorProvider = CreateInvestorProvider();

            var simulationCache = new SimulationCache();

            var simulationFactory = new SimulatorFactory(marketDataCache, simulationCache);
            var ratingService = new RatingService(
                marketDataCache,
                simulationFactory,
                investorProvider);
            var marketStatisticService = new MarketStatisticsProvider(marketDataCache);

            var strategyFactory = CreateStrategyFactory(
                marketDataCache,
                simulationCache,
                investorProvider,
                ratingService);

            var marketDataRepository = new Mock<IRepository<MarketData>>();
            var simulationResultsRepository = new Mock<IRepository<SimulationResult>>();

            _target = new ResultsProvider(
                marketDataCache,
                marketDataRepository.Object,
                ratingService,
                marketStatisticService,
                simulationResultsRepository.Object);

            _target.Initialise();

            var strategy = strategyFactory.Create(new RelativeStrengthParameters());
            var stateJson = File.ReadAllText(@"HolidayEffectSimulationState.json");
            var simulationState = JsonConvert.DeserializeObject<SimulationState[]>(stateJson);
            var resultsToAdd = new ConcurrentDictionary<IStrategy, SimulationState[]>();
            resultsToAdd.TryAdd(strategy, simulationState);

            _target.AddResults(investor, resultsToAdd);
        }

        [Test]
        public async Task PerformanceChart()
        {
            var path = Path.Combine(Environment.CurrentDirectory, "..", "..", "..", "ChartTests", "PerformanceChart.png");

            await _target.SaveChart(ResultsChart.Performance, path);

            Approvals.Verify(new FileInfo(path));
        }

        [Test]
        public async Task RelativeChart()
        {
            var path = Path.Combine(Environment.CurrentDirectory, "..", "..", "..", "ChartTests", "RelativeChart.png");

            await _target.SaveChart(ResultsChart.Relative, path);

            Approvals.Verify(new FileInfo(path));
        }

        [Test]
        public async Task SignalChart()
        {
            var path = Path.Combine(Environment.CurrentDirectory, "..", "..", "..", "ChartTests", "SignalChart.png");

            await _target.SaveChart(ResultsChart.Signal, path);

            Approvals.Verify(new FileInfo(path));
        }

        [Test]
        [UseApprovalSubdirectory("StatisticalTests")]
        public void PriceProbabilities()
        {
            var currentPrice = _target.CurrentPrice();
            var dailyProbability = _target.PriceProbability(Period.Day);
            var weeklyProbability = _target.PriceProbability(Period.Week);
            var monthlyProbability = _target.PriceProbability(Period.Month);
            var quarterlyProbability = _target.PriceProbability(Period.Quarter);
            var maximumProbability = _target.MaximumPriceProbability(Period.Day);

            Approvals.Verify(
@$"Current Price: {currentPrice:C2}
Daily: {dailyProbability:P2}
Weekly: {weeklyProbability:P2}
Monthly: {monthlyProbability:P2}
Quarterly: {quarterlyProbability:P2}
Maximum: (Price) {maximumProbability.Price:C2}, (Probability) {maximumProbability.Probability:P2}");
        }

        [Test]
        public async Task SummaryReport_Html()
        {
            var reportProvider = new ReportProvider(_target);

            var (investor, results) = _target.GetResults().First();
            var report = await reportProvider.GenerateReports(investor, results);

            Approvals.VerifyHtml(report.Summary.Body);
        }

        [Test]
        public async Task StrategyReport_Html()
        {
            var reportProvider = new ReportProvider(_target);

            var (investor, results) = _target.GetResults().First();
            var report = await reportProvider.GenerateReports(investor, results);

            Approvals.VerifyHtml(report.StrategyReports[0].Body);
        }

    }
}
