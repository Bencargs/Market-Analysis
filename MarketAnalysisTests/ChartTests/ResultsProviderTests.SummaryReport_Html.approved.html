<table style="height: 300px; width: 800px; margin-left: auto; margin-right: auto;">
    <tbody>
        <tr>
            <td><img style="float: left;" src="cid:logo" alt="CBC Market Analysis" width="240" height="200" /></td>
            <td style="text-align: right; float: right;">
                <p style="text-align: left;"><strong>CBC Market Analysis</strong><br />Mount Lawley<br />Perth W.A., 6050<br />Australia</p>
                <table>
                    <tbody>
                        <tr>
                            <td style="text-align: right;"><img src="cid:website" /></td>
                            <td style="text-align: left;"><a href="http://www.cbc.com.au">www.cbc.com.au</a></td>
                        </tr>
                        <tr>
                            <td style="text-align: right;"><img src="cid:phone" /></td>
                            <td style="text-align: left;">+61 450 307 000</td>
                        </tr>
                        <tr>
                            <td style="text-align: right;"><img src="cid:email" /></td>
                            <td style="text-align: left;"><a href="mailto:research@cbc.com.au">Research@cbc.com.au</a></td>
                        </tr>
                    </tbody>
                </table>
            </td>
        </tr>
        <tr>
            <td colspan="2"><hr />&nbsp;</td>
        </tr>
        <tr>
            <td>
                <table>
                    <tbody>
                        <tr>
                            <td style="text-align: left;"><span style="text-decoration: underline;">Market Recommendation</span>:</td>
                            <td style="text-align: left;">Hold</td>
                        </tr>
                        <tr>
                            <td style="text-align: left;"><span style="text-decoration: underline;">Current Price</span>:</td>
                            <td style="text-align: left;">$85.49</td>
                        </tr>
                        <tr>
                            <td style="text-align: left;"><span style="text-decoration: underline;">Stake:</span></td>
                            <td style="text-align: left;">$0.00 (100.00%)</td>
                        </tr>
                    </tbody>
                </table>
            </td>
            <td style="float: right;">
                <table>
                    <tbody>
                        <tr>
                            <td style="text-align: right;">Analysis Date:</td>
                            <td style="text-align: left;">10/12/2020</td>
                        </tr>
                        <tr>
                            <td style="text-align: right;">Investor Name:</td>
                            <td style="text-align: left;"></td>
                        </tr>
                        <tr>
                            <td style="text-align: right;">Investor Number:</td>
                            <td style="text-align: left;"></td>
                        </tr>
                    </tbody>
                </table>
            </td>
        </tr>
        <tr>
            <td colspan="2">
                &nbsp;
                <table style="margin-left: auto; margin-right: auto; white-space: nowrap;">
                    <tbody>
                        <tr>
                            <td><strong><u>Simulated Strategies</u></strong></td>
                            <td><strong><u>Recommendation</u></strong></td>
                            <td><strong><u>Profit YTD</u></strong></td>
                            <td><strong><u>Profit Total</u></strong></td>
                            <td align="right"><strong><u>Buy Count</u></strong></td>
                        </tr>
                        <tr><td>Relative Strength</td><td>Hold</td><td>$538.71</td><td>$30,456.61</td><td>114</td></tr>
                    </tbody>
                </table>
                &nbsp;
            </td>
        </tr>
        <tr>
            <td colspan="2"><span style="text-decoration: underline;">Profit to Date: $30,456.61</span>&nbsp;&nbsp;</td>
        </tr>
        <tr>
            <td colspan="2"><span style="text-decoration: underline;">Market Average: $34,223.25</span></td>
        </tr>
        <tr>
            <td colspan="2"><hr /></td>
        </tr>
		<tr>
            <td colspan="2">
                <table>
                    <tbody>
                        <tr>
                            <td style="text-align: left;"><span style="text-decoration: underline;">Extimated Order Execution:</span></td>
                        </tr>
                        <tr>
                            <td />
                            <td>Daily</td>
                            <td>Weekly</td>
                            <td>Monthly</td>
                            <td>Quarterly</td>
                        </tr>
                        <tr>
                            <td />
                            <td style="text-align: left;">46.43%&nbsp;</td>
                            <td style="text-align: left;">53.33%&nbsp;</td>
                            <td style="text-align: left;">42.86%&nbsp;</td>
                            <td style="text-align: left;">45.83%&nbsp;</td>
                        </tr>
                    </tbody>
                </table>
            </td>
		<tr>
            <td colspan="2"><hr />&nbsp;</td>
        </tr>
        <tr>
            <td colspan="2">
                <p style="text-align: center;"><strong>CBC Market Analytics<br /></strong>(ABN 00 000 000 000)</p>
                <p style="text-align: center;"><a href="mailto:research@cbc.com.au">Research@cbc.com.au</a></p>
                &nbsp;
            </td>
        </tr>
    </tbody>
</table>